### PR TITLE
`bundle gem` check if git is installed

### PR DESCRIPTION
**Pull request by Shekharrajak**
Fixes https://github.com/bundler/bundler/issues/5458

### CODE REVIEW

File: lib/bundler/cli/gem.rb [hunk 1]
```diff
@@ -29,6 +29,11 @@ def run
       constant_name = name.gsub(/-[_-]*(?![_-]|$)/) { "::" }.gsub(/([_-]+|(::)|^)(.|$)/) { $2.to_s + $3.upcase }
       constant_array = constant_name.split("::")
 
+      if !Bundler.git_present?
```
Comment (hmistry) [inline lib/bundler/cli/gem.rb]: use `unless` instead of `if !`

File: lib/bundler/cli/gem.rb [hunk 2]
```diff
@@ -29,6 +29,11 @@ def run
...
+        Bundler.ui.error "You need to install git to be able to use bundle gem."
```
Comment (colby-swandale) [inline lib/bundler/cli/gem.rb]: Running `bundle gem` without git installed should work as expected for the user. We should just be setting empty values for `author`, `username` and `email` if we cannot source it from git.

**Comment by colby-swandale** [comment]
We also need a spec testing this behaviour.

File: lib/bundler/cli/gem.rb (hunk 2 above)
Comment (hmistry) [inline lib/bundler/cli/gem.rb]: @colby-swandale if that's the case, then this is not the right fix. I see your point in allowing it but it's not clear to me from bundler doc's if `bundle gem` is allowed to work without `git` or it requires `git`. Let's clarify this so the correct fix is implemented.

@indirect @segiddins any thoughts on git requirement for `bundle gem`?

**Comment by Shekharrajak** [comment]
Thanks @colby-swandale  @hmistry for review. I have edited the PR. Now `git_author_name`, `git _username`, `git _user_email` will be set to empty string when git is not installed. 

So in the created `mygem.gemspec` , these lines will be present(similar as when git author, email is absent)

```
  spec.authors       = ["TODO: Write your name"]
  spec.email         = ["TODO: Write your email address"]
```

and in `README.md` file `[USERNAME]` will be present instead of github username. (Means this : `https://github.com/[USERNAME]/mygem` ).

I am still finding a way to write spec for this.

File: lib/bundler/cli/gem.rb (hunk 2 above)
Comment (Shekharrajak) [inline lib/bundler/cli/gem.rb]: `git ` is needed only for fetching `username` `user_email` `git_author_name`, to write them in generated` mygem.gemspec` file and `README.md` file, which can be edited by the user afterward. 

So now I think it must not stop generating files (When git is not installed then just replace `username`, `email` with `["TODO: Write your name/email"]` in generated `mygem.gemspec` file)

File: spec/commands/newgem_spec.rb [hunk 3]
```diff
@@ -144,6 +144,23 @@ def gem_skeleton_assertions(gem_name)
...
+        `rm "#{ENV["GIT_CONFIG"]}"` if File.exist?(ENV["GIT_CONFIG"])
+        reset!
+        in_app_root
+        bundle "gem #{gem_name}"
+        remove_push_guard(gem_name)
+      end
+
+      it "contribute URL set to [USERNAME]" do
+        expect(bundled_app("test_gem/README.md").read).to include("[USERNAME]")
+        expect(bundled_app("test_gem/README.md").read).not_to include("github.com/bundleuser")
```
Comment (colby-swandale) [inline spec/commands/newgem_spec.rb]: You probably will want to test the `gemspec` as well to make sure we're setting sensible defaults.

File: spec/commands/newgem_spec.rb [hunk 4]
```diff
@@ -144,6 +144,23 @@ def gem_skeleton_assertions(gem_name)
...
+      end
+
+      it_should_behave_like "git config is absent"
```
Comment (Shekharrajak) [inline spec/commands/newgem_spec.rb]: @colby-swandale this line is using [shared example](https://github.com/bundler/bundler/pull/5466/files/4dec0e6f2c980c2b21fc96c987fbda5c6c351093#diff-3fbe0dfd8b4d505d057982532848c10eR66) to check the `gemspec` file. It must put `"TODO: Write your name"` , `"TODO: Write your email` in the place of author name, email (same as when git config is unset).

File: spec/commands/newgem_spec.rb (hunk 3 above)
Comment (Shekharrajak) [inline spec/commands/newgem_spec.rb]: It is in [line 162](https://github.com/bundler/bundler/pull/5466#discussion_r102673468)

**Comment by Shekharrajak** [comment]
Thanks @hmistry  for review.

File: spec/commands/newgem_spec.rb [hunk 5]
```diff
@@ -144,6 +144,23 @@ def gem_skeleton_assertions(gem_name)
         expect(bundled_app("test_gem/README.md").read).not_to include("github.com/bundleuser")
       end
     end
+
+    context "git is not installed in the system" do
+      before do
+        `rm "#{ENV["GIT_CONFIG"]}"` if File.exist?(ENV["GIT_CONFIG"])
```
Comment (segiddins) [inline spec/commands/newgem_spec.rb]: please use `FileUtils` to `rm`

**Comment by segiddins** [comment]
@colby-swandale want to re-review this?

**Comment by colby-swandale** [comment]
Sorry, i didn't mean to give the impression to not merge this PR without my requested changes, only men't as suggestions. I'll keep that in mind in the future @segiddins

File: spec/commands/newgem_spec.rb (hunk 5 above)
Comment (segiddins) [inline spec/commands/newgem_spec.rb]: This is the one blocker to merging

Comment (Shekharrajak) [inline spec/commands/newgem_spec.rb]: I will edit this PR soon.

**Comment by segiddins** [comment]
@bundlerbot r+

**Pull request merged**